sandboxify.c

```diff
@@ -39,75 +41,127 @@ int main(int argc, char **argv)
+			return EXIT_FAILURE;
+		}
+
+	} else {
+		int status = 0;
+		pid_t waitp;
+
+		/* Ensure we are killed is the parent exits */
+		if (prctl(PR_SET_PDEATHSIG, SIGKILL, 0, 0, 0) < 0) {
+			perror("failed to set pdeathsig bit for the child");
+			return EXIT_FAILURE;
+		}
+		/* Attach to the parent and interrupt it. No signal used. */
+		if (ptrace(PTRACE_SEIZE, parent, NULL, NULL)) {
+			perror("failed to seize parent");
+			return EXIT_FAILURE;
+		}
+		if (ptrace(PTRACE_INTERRUPT, parent, NULL, NULL)) {
+			perror("failed to interrupt parent");
+			return EXIT_FAILURE;
+		}
+
+		waitp = waitpid(parent, &status, 0);
+
+		/* this should be executed after the first ptrace-stop from the parent
+		 * which PTRACE_INTERRUPT above ensures.
+		 * */
+		if (ptrace(PTRACE_SETOPTIONS, parent, NULL, (TRACE_OPTS | PTRACE_O_SUSPEND_SECCOMP | PTRACE_O_EXITKILL))) {
+			perror("failed to suspend seccomp filters for the parent");
+			/* PTRACE_O_EXITKILL was not yet set, but it should catch this
+			 * child's exit. */
 			return EXIT_FAILURE;
 		}
-		
+
 		while (waitp > 0) {
 			if (-1 == waitp) {
 				perror("waitpid failed");
 				return EXIT_FAILURE;
 			}
+			/* Note we probably will never see either of the EXITED/SIGNALED
+			 * cases because of PR_SET_PDEATHSIG.  */
 			if (WIFEXITED(status)) {
-				fprintf(stderr, "child exited with code %d\n", WEXITSTATUS(status));
+				fprintf(stderr, "parent exited with code %d\n", WEXITSTATUS(status));
 				return EXIT_SUCCESS;
 			}
 			if (WIFSIGNALED(status)) {
-				fprintf(stderr, "child was terminated by signal %d\n", WTERMSIG(status));
+				fprintf(stderr, "parent was terminated by signal %d\n", WTERMSIG(status));
 				return EXIT_SUCCESS;
 			}
+
 			if (WIFSTOPPED(status)) {
 				if ((SIGTRAP | (PTRACE_EVENT_EXEC << 8)) == (status >> 8)) {
-					/* child is stopped before returning from exec */
+					/* parent is stopped before returning from exec. */
 					/* re-enable seccomp filters and detach */
-					if (ptrace(PTRACE_SETOPTIONS, child, NULL, TRACE_OPTS)) {
-						perror("failed to re-enable seccomp filters for the child");
+					if (ptrace(PTRACE_SETOPTIONS, parent, NULL, TRACE_OPTS)) {
+						perror("failed to re-enable seccomp filters for the parent");
 						return EXIT_FAILURE;
 					}
-					if (ptrace(PTRACE_DETACH, child, NULL, NULL)) {
-						perror("failed to detach from the child");
+					if (ptrace(PTRACE_DETACH, parent, NULL, NULL)) {
+						perror("failed to detach from the parent");
 						return EXIT_FAILURE;
 					}
 					return EXIT_SUCCESS;
 				}
+				if ((SIGTRAP | (PTRACE_EVENT_STOP << 8)) == (status >> 8)) {
+					/* parent is stopped before running a syscall. */
+					struct user_regs_struct regs;
 
-				if (ptrace(PTRACE_CONT, child, NULL, NULL)) {
-					perror("failed to resume the child");
+					if (ptrace(PTRACE_GETREGS, parent, 0, &regs)) {
+						perror("failed to read regs from the parent");
+						return EXIT_FAILURE;
+					}
+					if (regs.orig_rax == 61) {
```

Comment:
Does this make the code `x86` specific? We'd really like to stay arch independent.

Reply:
it does, unfortunately. 
As far as I can tell, overriding system calls is not portable (maybe someone with more PTRACE-foo can chime in?)

If we remove the blocking of the parent on a wait call, what do you think of a simple pipe shared between parent and child?
Parent blocks on the pipe `read()` until it is PTRACEd.
If child exits, we get an event.
Once PTRACE is setup, we can close the pipe in the child which will wake the parent and close the pipe in the parent before setting up SECCOMP & `execvp`.

What do you think?

Reply:
The reason I am avoiding the `raise(SIGSTOP)` is that it is visible in the parent of `sandboxify` (since PTRACE is not yet setup to suppress it).

Reply:
pipe seems reasonable

it seems the kernel itself uses a similar approach for some tests https://github.com/torvalds/linux/blob/1566feea450cb0ffcdb47daf723f62640c9f616a/tools/testing/selftests/seccomp/seccomp_bpf.c#L1454-L1469